---
title: "01 iPSC"
author: "Pierre-Luc Germain"
date: "7/15/2021"
output:
    html_document:
        toc: true
        toc_float: true
        theme: 'cerulean'
        highlight: 'tango'
        code_folding: hide
        df_print: paged
---

<style>
  .main-container {
    max-width: 1500px !important;
  }
</style>

=====================
Pierre-Luc Germain - `r format(Sys.time(), '%d %B, %Y')`
<br><br><br>

```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SEtools)
  library(ComplexHeatmap)
  library(ggplot2)
  library(cowplot)
})
source("../misc.R")
source("../colors.R")
```

```{r, fig.width=3.5, fig.height=4.5}
se <- toSymbol(readRDS("../clean/Neurons/RNA/Neurons.RNA.merged.SE.rds"))
se <- se[,se$genotype!="AtWBS"]
se$genotype <- droplevels(se$genotype)
levels(se$genotype) <- c("CTL","WBS","7Dup")
se$genotype <- factor(as.character(se$genotype), c("WBS","CTL","7Dup"))
se$Dataset <- gsub("patientDerived","patient-\nderived",se$Dataset)
se$Dataset <- factor(se$Dataset, c("patient-\nderived","isogenic"))
WBScols = c(WBS="#ff2600", CTL="#c0c0c0", "7Dup"="#0433ff")
pdf("fig1_wbs.pdf", width=3.5, height=4)
draw(SEtools::sechm(se, WBSg1, assay="log2FC", sortRowsOn=NULL, cluster_rows=FALSE, gaps_at="Dataset",
      anno_columns=c("genotype"), breaks=0.99, row_names_gp=gpar(fontsize=10), anno_colors=list(genotype=WBScols), use_raster=TRUE, row_title="WBSCR genes ordered by genomic position",
      row_title_gp=gpar(fontsize=11)), merge_legends=TRUE)
dev.off()
```

# trisomy

```{r, fig.width=9, fig.height=3}
iso <- readRDS("../clean/Neurons/RNA/Neurons.RNA.isogenic.DEA.rds")
gr <- rtracklayer::import("~/bioinfo/annotation/human/gencode.v29.annotation.gtf")
gr <- gr[gr$type=="gene"]
gch <- setNames(as.factor(GenomicRanges::seqnames(gr)), gr$gene_name)
iso$chr <- gch[gsub("ENSG[0-9]+\\.","",row.names(iso))]
iso <- iso[iso$baseMean>10,]
d2 <- data.frame(logFC=c(iso$logFC.WBS, iso$logFC.DUP), chr=rep(iso$chr,2), condition=rep(c("WBS","7Dup"),each=nrow(iso)))
levels(d2$chr) <- gsub("chr","",levels(d2$chr))
d2 <- d2[d2$chr %in% 1:22,]
p <- ggplot(d2, aes(chr, logFC)) + 
  geom_violin(fill="grey", scale="width") + stat_summary(fun="mean", geom="point") +
  facet_wrap(~condition, nrow=2) + ylim(-1,1) + geom_hline(yintercept = 0, linetype="dashed") +
  xlab("Chromosome") + ylab("RNAseq log(fold-change)")
pdf("logFC_byChr.pdf", width=8, height=3)
p
dev.off()
```

