---
title: "Neurons 2"
output:
    html_document:
        toc: true
        toc_float: true
        theme: 'cerulean'
        highlight: 'tango'
        code_folding: hide
        df_print: paged
---

<style>
  .main-container {
    max-width: 1500px !important;
  }
</style>

=====================
Pierre-Luc Germain - `r format(Sys.time(), '%d %B, %Y')`
<br><br><br>


```{r, message=FALSE}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SEtools)
  library(ggplot2)
  library(ggrepel)
  library(cowplot)
  library(grid)
  library(ComplexHeatmap)
  library(patchwork)
})
source("../misc.R")
source("../misc/enrichmentAnalysis.R")
source("../misc/crossLayer.R")
source("../colors.R")
asd <- readRDS("../extData/ID_ASD_genes.rds")
```

```{r}
gsets <- getMsigSets()
gsets <- gsets[grep("C5:GO", names(gsets), fixed=TRUE)]
```


```{r}
SEs <- list( RNA=toSymbol(readRDS("../clean/Neurons/RNA/Neurons.RNA.merged.SE.rds")),
             #TE=toSymbol(readRDS("../clean/Neurons/TE/Neurons.TE.isogenic.SE.rds")),
             Protein=readRDS("../clean/Neurons/Protein/Neurons.Protein.merged.SE.rds") )
#SEs$TE$Dataset <- "isogenic"
SEs <- lapply(SEs, FUN=function(x){ x$system <- x$Dataset; x })
SEs <- lapply(SEs, FUN=function(x){
  x <- x[,which(x$genotype!="AtWBS")]
  x$genotype <- droplevels(x$genotype)
  x
})
#SEs$TE <- log2FC(SEs$TE, "logTE", SEs$TE$genotype=="CTRL")

dea <- list( RNA=readRDS("../clean/Neurons/RNA/Neurons.RNA.merged.DEA.rds"),
             #TE=readRDS("../clean/Neurons/TE/Neurons.TE.isogenic.DEA.rds"),
             Protein=readRDS("../clean/Neurons/Protein/Neurons.Protein.merged.DEA.rds"))
dea <- lapply(dea, toSymbol)
SEs$Protein <- SEs$Protein[rowSums(is.na(assays(SEs$Protein)$intensity))<=3,]
```


```{r}
degs <- union(getDEGs.any(dea$RNA,logFC.thres=log2(1.2), FDR.thres=0.05),
              getDEGs.any(dea$Protein,logFC.thres=log2(1.2)) )

SEs <- lapply(SEs, FUN=function(x) x[,order(factor(x$genotype,c("WBS","CTRL","DUP")), x$system)])
sem <- SEtools::mergeSEs(SEs, "log2FC", do.scale=FALSE, commonOnly=TRUE)
sem$layer <- sem$Dataset <- factor(sem$Dataset, c("RNA","TE","Protein"))
metadata(sem)$anno_colors <- list()
sem2 <- SEtools::mergeSEs(SEs, "log2FC", do.scale=FALSE, commonOnly=TRUE)
sem2$layer <- sem2$Dataset <- factor(sem2$Dataset, c("RNA","Protein"))
levels(sem$genotype)[1] <- levels(sem2$genotype)[1] <- "CTL"


degs <- intersect(degs, row.names(sem))
# restrict to genes going in the same direction in the two protein datasets
# deas <- list(iso=readRDS("../clean/Neurons/Protein/Neurons.Protein.isogenic.DEA.rds"),
#              pD=readRDS("../clean/Neurons/Protein/Neurons.Protein.patientDerived.DEA.rds"))
# deas <- lapply(deas, toSymbol)
# m <- merge(deas$iso, deas$pD, by="row.names")
# m <- m[sign(m$logFC.x)==sign(m$logFC.y) & ((m$PValue.x<0.05 & m$PValue.y<0.05) & ,]
# degs <- intersect(degs,m$Row.names)
```

## Clustering of RNA+Prot DEGs

```{r}
binarize <- function(i, obj, lfc.th=0.2, scaled=TRUE){
  if(scaled){
    x <- t(scale(t(assays(obj)$log2FC[degs,]), center=FALSE))[,i,drop=FALSE]
  }else{
    x <- assays(obj)$log2FC[degs,i,drop=FALSE]
  }
  y <- sign(rowMedians(x))*(abs(rowMedians(x))>lfc.th)
  w <- apply(sign(x),1,FUN=function(x) length(unique(x))>1)
  #w <- apply(sign(x),1,FUN=function(x) (max(table(x))/length(x))<0.8)
  y[w] <- y[w]/2
  y
}
b1 <- sapply(split(seq_len(ncol(SEs$Protein)), paste(SEs$Protein$genotype)), obj=SEs$Protein, FUN=binarize)
b1 <- b1[,grep("AtWBS|CTRL",colnames(b1),invert=TRUE)]
b2 <- sapply(split(seq_len(ncol(SEs$RNA)), paste(SEs$RNA$genotype)), obj=SEs$RNA, FUN=binarize)
b2 <- b2[,grep("AtWBS|CTRL",colnames(b2),invert=TRUE)]
b <- cbind(b1,b2)
row.names(b) <- degs
b2 <- b[rowSums(is.na(b))==0,]
set.seed(123)
k <- kmeans(b2, 8, nstart=3)

rowData(sem)$cluster <- 0L
rowData(sem)[row.names(b2),"cluster"] <- k$cluster
sechm(sem, degs, gaps_row="cluster", do.scale=TRUE)
```

```{r}
# conv <- "converging\npost-\ntranslationally"
fo <- "forwarded to\nthe proteome,\nopposite between\ngenotypes"
# fs <- "forwarded to\nthe proteome,\nshared between\ngenotypes"
# cl2 <- c("none",conv,fo,conv,conv,conv,fo,conv,"buffered in DUP")
cl2 <- c("none","buffered\nin WBS","buffered\nin DUP","buffered\nin WBS","buffered\nin DUP","buffered\nin both","buffered\nin DUP","buffered\nin DUP",fo)
```



```{r}
rowData(sem)$cluster2 <- factor(cl2[1L+rowData(sem)$cluster])
assays(sem)$scaledLFC <- sechm:::safescale(assays(sem)$log2FC, center=FALSE, byRow=TRUE)
h <- sechm(sem, degs, do.scale=FALSE, top_annotation=c("system", "genotype"), assayName="scaledLFC",
      breaks=0.99, gaps_at="Dataset", gaps_row="cluster2", row_title_rot=0, row_title_gp=gpar(fontsize=10))
```

Some nice examples:

```{r}
#cs3 <- lapply(cs2, FUN=intersect, y=rna.degs2)
seldegs <- list(
  buffered=c("S100A11", "SCML2", "TPM2", "LMANL2", "CNRIP1"),
  "forwarded\nto the\nproteome"=c("MAPT", "CLIP2", "SNCA", "MAP2", "DPYSL3", "DPYSL3")
)
sechm(sem, seldegs, do.scale=FALSE, top_annotation=c("system", "genotype"), assayName="log2FC",
      breaks=0.975, gaps_at="Dataset", gaps_row="cluster3", row_title_rot=0, row_title_gp=gpar(fontsize=10))

pdf("examples_buff_forw.pdf", width=6, height=2.1)
draw(sechm(sem, seldegs, do.scale=FALSE, top_annotation=c("system", "genotype"), assayName="log2FC", use_raster=TRUE,
      breaks=0.975, gaps_at="Dataset", row_title_rot=0, row_title_gp=gpar(fontsize=10), row_names_gp=gpar(fontsize=10)),
     merge=TRUE)
dev.off()
```



```{r}
cs2 <- split(degs, droplevels(rowData(sem)[degs,"cluster2"]))
go2 <- lapply(cs2, FUN=function(x) goseq.enrichment(row.names(sem), x))
#go3 <- lapply(cs2, FUN=function(x) topgo(row.names(sem), x))
```

```{r}
mdea <- merge(dea$RNA, dea$Protein, by="row.names", suffix=c(".RNA", ".Protein"))
mdea2 <- mdea[which(mdea$Row.names %in% cs2[[grep("forwarded",names(cs2))]]),]
mods <- lapply(paste0("logFC", c("",".WBS",".DUP")), FUN=function(x){
  MASS::rlm(mdea2[[paste0(x,".Protein")]]~0+mdea2[[paste0(x,".RNA")]])
})
lapply(mods, summary)
```


```{r, fig.width=10, fig.height=5}
gofo <- go2[[grep("oppo",names(go2))]]
set.seed(1234)
gofo <- clusterGO(gofo[gofo$FDR<=0.05,], k=5, fontsize = 2.5) + theme_cowplot(10.5) + 
  theme(title=element_text(size=10.5)) + 
  ggtitle("Forwarded to the proteome, opposite between genotypes")
set.seed(1234)
gobd <- clusterGO(go2[[grep("DUP",names(go2))]], k=2:5, fontsize = 2.5) + 
  theme_cowplot(10.5) + theme(title=element_text(size=10.5)) + 
  ggtitle("Buffered in DUP")
plot_grid(
  grid.grabExpr(draw(h, merge=TRUE)),
  (gofo / gobd), nrow=1, scale=0.95, rel_widths = c(4,5)
)
```




## Buffering coefficient

```{r}
m <- lapply(list("Regression\non copynumber"=NULL, WBS="WBS", DUP="DUP"), FUN=function(x){
  rna <- subset_mDEA(dea$RNA, x)
  prot <- subset_mDEA(dea$Protein, x)
  degs <- union(getDEGs(rna, log2(1.3), 0.01), getDEGs(prot))
  degs <- intersect(degs, intersect(row.names(prot),row.names(rna)))
  m <- merge(rna, prot, by="row.names", suffix=c(".RNA", ".Protein"))
  m$delta.logFC <- m$logFC.Protein-m$logFC.RNA
  m$buffering.prop <- ifelse(m$logFC.RNA>0,
    (2^(m$logFC.RNA)-2^(m$logFC.Protein))/2^(m$logFC.RNA),
    (2^abs(m$logFC.RNA)-2^(-m$logFC.Protein))/2^abs(m$logFC.RNA))
  m$buffering.prop.capped <- pmax(pmin(1,m$buffering.prop), -1)
  m$sig <- m$Row.names %in% degs
  m$buffering.prop.capped <- ifelse(sign(m$logFC.RNA)!=sign(m$logFC.Protein) & m$sig,
                                    1, m$buffering.prop.capped)
  m
})
m2 <- merge(m$WBS, m$DUP, by="Row.names", suffix=c(".WBS",".DUP"))
m2$significance <- factor(m2$sig.DUP+2*m2$sig.WBS, 0:3, c("copynumber", "DUP", "WBS", "both"))
m2$meanLog10FDR <- rowMeans(-log10(as.matrix(m2[,grep("FDR", colnames(m2))])))
m2$strongestAbsLFC <- matrixStats::rowMaxs(abs(as.matrix(m2[,grep("logFC\\.RNA", colnames(m2))])))
m3 <- m2[which(m2$Row.names %in% degs),]
ggplot(m3, aes(delta.logFC.WBS, delta.logFC.DUP, colour=logFC.RNA.DUP-logFC.RNA.WBS)) + 
  coord_cartesian(xlim=c(-8,5), ylim=c(-8,5)) + geom_vline(xintercept=0, linetype="dashed") +
  geom_hline(yintercept=0, linetype="dashed") + geom_point() + 
  scale_colour_gradient2(breaks=c(-3, 0, 3), low = "darkred", mid="grey", high = "darkblue")
ggplot(m3, aes(buffering.prop.capped.WBS, buffering.prop.capped.DUP, colour=strongestAbsLFC)) + 
  geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + 
  geom_point() + scale_color_viridis_c(limits=c(0,2))
buffInBoth <- as.character(m2$Row.names[which(m2$buffering.prop.capped.DUP>0.5 & m2$buffering.prop.capped.WBS>0.5)])
sechm(sem, buffInBoth, do.scale=FALSE, top_annotation=c("system", "genotype"), breaks=0.99, gaps_at="Dataset")
d <- meltSE(sem, intersect(buffInBoth, row.names(sem)))
d <- d[d$genotype != "AtWBS" & d$Dataset!="TE",]
ggplot(d, aes(genotype, log2FC, fill=system)) + geom_boxplot() + facet_wrap(~Dataset)
```

```{r}
rowData(sem)$buff.WBS <- rowData(sem)$buff.DUP <- NA
rowData(sem)[as.character(m3$Row.names), "buff.DUP"] <- m3$buffering.prop.capped.DUP
rowData(sem)[as.character(m3$Row.names), "buff.WBS"] <- m3$buffering.prop.capped.WBS
rowData(sem)$buff.7Dup <- rowData(sem)$buff.DUP
levels(sem$genotype)[3] <- "7Dup"
sechm(sem[,sem$genotype!="AtWBS"], degs, do.scale=FALSE, top_annotation=c("system", "genotype"),
           breaks=0.99, gaps_at="Dataset",gaps_row="cluster2", row_title_rot=0, row_title_gp=gpar(fontsize=10), left_annotation=c("buff.7Dup", "buff.WBS"),
      anno_colors = list(buff.DUP=RColorBrewer::brewer.pal(10, "RdBu")))

sebuf <- SummarizedExperiment(list(buffering=as.matrix(rowData(sem)[,c("buff.WBS","buff.7Dup")])),
                              rowData=rowData(sem)[,"cluster2", drop=FALSE],
                              colData=data.frame(genotype=c("WBS","7Dup")))
metadata(sebuf)$hmcols <- rev(RColorBrewer::brewer.pal(11, "RdBu"))
metadata(sebuf)$anno_colors <- list(genotype=metadata(sem)$anno_colors$genotype)

degs2 <- unlist(lapply(cs2, FUN=function(degs){
  row.names(sortRows(cbind(3.5*assay(sebuf)[degs,], assay(sem)[degs,])))
}))
br <- c(-1,seq(from=-0.7,to=-0.1,length.out=4),0,seq(from=0.1,to=0.7,length.out=4),1)
h <- sechm(sem[,sem$Dataset=="RNA"], degs2, do.scale=FALSE, top_annotation=c("system", "genotype"),
           cluster_rows = FALSE, sortRowsOn = NULL, gaps_row = "cluster2", assayName = "scaledLFC",
           row_title_rot=0, row_title_gp=gpar(fontsize=10), 
        breaks=0.99, gaps_at="Dataset", use_raster = TRUE, isMult = TRUE, name="RNA scaledLFC") +
  sechm(sebuf, degs2, cluster_rows = FALSE, sortRowsOn = NULL, do.scale=FALSE, gaps_row = "cluster2", width=unit(1, "cm"),
      heatmap_legend_param=list(at=c(-1,0,1), labels=c("amplified","forwarded","buffered")),
      breaks=br, use_raster=TRUE) +
  sechm(sem[,sem$Dataset=="Protein"], degs2, do.scale=FALSE, top_annotation=c("system", "genotype"), assayName = "scaledLFC",
        breaks=0.99, gaps_at="Dataset", cluster_rows = FALSE, sortRowsOn = NULL, use_raster = TRUE)
```

```{r, fig.width=10, fig.height=5}
pp <- plot_grid(
  grid.grabExpr(draw(h, merge=TRUE)),
  (gofo / gobd), nrow=1, scale=0.95, rel_widths = c(4.5,5)
)
pdf("translation_clusters.pdf", width=10, height=5)
pp
dev.off()
```


```{r}
# bufw <- fgsea::fgseaMultilevel(gsets, setNames(assay(sebuf)[degs2,1],degs2))
# bufw <- as.data.frame(bufw)[order(bufw$pval, abs(bufw$NES)),]

d <- meltSE(sebuf, degs2)
d$sample <- gsub("buff\\.","",d$sample)
ggplot(d, aes(buffering, colour=sample)) + geom_vline(xintercept=0, linetype="dashed") +
  geom_density(size=1.5) + coord_cartesian(xlim=c(-0.7,0.7)) +
  scale_x_continuous(breaks=c(-0.7,0,0.7), labels=c("amplified","forwarded","buffered")) +
  scale_colour_manual(values=WBScols) +
  labs(x="Buffering coefficient", colour="genotype")
```


```{r, fig.width=10, fig.height=6}
cs2 <- split(degs, droplevels(rowData(sem)[degs,"cluster2"]))
go2 <- lapply(cs2, FUN=function(x) goseq.enrichment(row.names(sem), x))
#go3 <- lapply(cs2, FUN=function(x) topgo(row.names(sem), x))
go.buf <- goseq.enrichment(row.names(sem), unlist(cs2[grep("buffered",names(cs2))]))
set.seed(1234)
gobuf <- clusterGO(go.buf,k=4, fontsize=2.5) + theme_cowplot(10.5) + 
  theme(title=element_text(size=10.5)) + 
  ggtitle("Buffered (in any genotype)")
set.seed(1234)
gofo <- clusterGO(subset(go2[[grep("fo",names(go2))]], FDR<=0.05), k=6, fontsize=2.5) + theme_cowplot(10.5) + 
  theme(title=element_text(size=10.5)) + 
  ggtitle("Forwarded to the proteome, opposite between genotypes")
gofo
```

```{r, eval=FALSE}
sechm(sem, intersect(intersect(getDEGs(dea$Protein, FDR.thres=0.25), cs2$`forwarded
opposite`), c(asd$SPARK,asd$MSSNG,unlist(asd$sfari))), do.scale=FALSE, top_annotation =c("system", "genotype"), breaks=0.99, show_rownames=TRUE)
```


# logFC scatterplots

```{r, eval=FALSE}
wbs <- crossLayer(lapply(dea[c(1,3,2)], suffix="WBS", FUN=subset_mDEA))
wbs$gene <- row.names(wbs)
wbs <- wbs[intersect(row.names(wbs), degs),]
wbs$cluster <- rowData(sem)[row.names(wbs),"cluster"]
dup <- crossLayer(lapply(dea[c(1,3,2)], suffix="DUP", FUN=subset_mDEA))
dup$cluster <- rowData(sem)[row.names(dup),"cluster"]
dup$gene <- row.names(dup)
dup <- dup[intersect(row.names(dup), degs),]
wdup <- which(dup$PValue.TE<0.01 & (dup$translational | dup$buffered))
pd <- plot3Layers(dup) + xlim(c(-1.5,2.)) + ylim(c(-2,3.5)) + geom_text_repel(data=dup[wdup,,drop=FALSE], aes(label=gene)) + scale_colour_gradient2(low="blue", mid="lightgrey", high="red") + ggtitle("Dup")
wwbs <- which(wbs$PValue.TE<0.01 & (wbs$translational | wbs$buffered))
pw <- plot3Layers(wbs) + geom_text_repel(data=wbs[wwbs,,drop=FALSE], aes(label=gene)) + scale_colour_gradient2(low="blue", mid="lightgrey", high="red") + ggtitle("WBS") + xlim(c(-1.5,2.5))
```

```{r, fig.width=10, fig.height=9}
# h <- sechm(sem, degs, breaks=0.99, top_annotation=c("genotype","system"), gaps_at="Dataset", mark=union(wbs$gene[wwbs],dup$gene[wdup]), sortRowsOn=which(sem$Dataset!="TE"), do.scale=TRUE, sort.method="R2E", gaps_row="cluster2", use_raster=TRUE)
# pgo <- plot_grid(gofo + theme(legend.position="none") + ggtitle("Forwarded opposite"), 
#                           gobuf + theme(legend.position="none") + ggtitle("Buffered in DUP"), 
#                           nrow=2, labels=c("D","E")) 
# p <- plot_grid(pw + labs(size="-log10\nTE.pval"), pd + labs(size="-log10\nTE pval"),
#                 grid.grabExpr(draw(h, merge_legends=TRUE)),
#                 pgo, labels=c("A","B","C",""), nrow=2, rel_heights=c(3,6))
# p
```



# TOP

```{r, fig.width=7, fig.height=3}
rna <- toSymbol(readRDS("../clean/Neurons/RNA/Neurons.RNA.merged.DEA.rds"))
prot <- toSymbol(readRDS("../clean/Neurons/Protein/Neurons.Protein.merged.DEA.rds"))
rpf <- readRDS("../clean/Neurons/RPF/Neurons.RPF.isogenic.DEA.rds")
te <- readRDS("../clean/Neurons/TE/Neurons.TE.isogenic.DEA.rds")
top <- readLines("../extData/TOP.genes")

layers <- list("RNA"=rna,"TE"=te,"RPF"=rpf, "Protein"=prot)
conv <- c("other","TOP")
names(conds) <- conds <- c("WBS","DUP")


# pl <- unlist(lapply(conds, FUN=function(cond){
#   lapply(names(layers), FUN=function(layer){
#     x <- layers[[layer]]
#     x$group <- conv[1+(row.names(x) %in% top)]
#     p <- cdfplot(x[[paste("logFC",cond,sep=".")]], x$group, size=1.5) + xlim(-1,1) + 
#       xlab(paste(layer,"logFC in", cond))
#     if(layer=="Protein" & cond=="DUP"){
#       p <- p + theme(legend.position=c(.65,.2))
#     }else{
#       p <- p + theme(legend.position="none")
#     }
#     p
#   })
# }), recursive=FALSE)
# plot_grid(plotlist=pl, nrow=2)


d <- dplyr::bind_rows(lapply(layers, FUN=function(x){
  dplyr::bind_rows(lapply(conds, FUN=function(cond){
    ll <- split(x[[paste("logFC",cond,sep=".")]], conv[1+(row.names(x) %in% top)])
    dplyr::bind_rows(lapply(ll, FUN=function(x){
      x <- x[!is.na(x) & !is.infinite(x)]
      data.frame( y=seq_along(x)/length(x),
                  logFC=sort(x) )
    }), .id="Geneset")
  }), .id="condition")
}), .id="layer")
d$condition <- factor(d$condition, names(conds))
levels(d$condition)[2] <- "7Dup"
d$layer <- factor(d$layer, names(layers))
p <- ggplot(d, aes(x,y,colour=Genesets)) +
    geom_vline(xintercept=0, linetype="dashed") + geom_line()
pcdf <- ggplot(d, aes(logFC, y, colour=Geneset)) + 
    geom_vline(xintercept=0, linetype="dashed") + geom_line(size=1.3) + 
  facet_grid(condition~layer) + ylab("Cumulative proportion") +
  theme_cowplot(10.5) + theme(legend.position="bottom") + 
  scale_x_continuous(breaks=c(-1,0,1)) + 
  scale_y_continuous(breaks=c(0,1)) + coord_cartesian(xlim=c(-1,1))
  
pp <- plot_grid(pcdf, labels="F")
pp
ggsave("cdfs.pdf", pp, width=10, height=3)
```

```{r}
dp <- dplyr::bind_rows(lapply(split(d, d[,c("layer","condition")]), FUN=function(d2){
  slfc <- split(d2$logFC, d2$Geneset)
  p <- paste0("p~",format(ks.test(slfc[[1]], slfc[[2]])$p.value, digits=2))
  data.frame(layer=d2$layer[1], condition=d2$condition[1], p=p)
}))
pcdf <- pcdf + geom_text(data=dp, aes(label=p), x=-.98, y=0.9, colour="black", size=2.8, hjust=0)
```


```{r, fig.width=10, fig.height=7}
pp <- plot_grid(
  plot_grid(
    grid.grabExpr(draw(h, merge=TRUE)),
    plot_grid(gofo,gobuf,nrow=2,labels=c("B","C")), nrow=1, scale=0.95,
    rel_widths = c(4.5,5), labels=c("A",NA)),
  plot_grid(pcdf, nrow=1, scale=0.95, labels=c("D")),
  nrow=2, rel_heights=c(2,1)
)
pdf("translation_omics_figure2.pdf", width=10, height=7)
pp
dev.off()
```
